internal/controller/apiextensions/composite/composition_ptf.go

```diff
@@ -0,0 +1,982 @@
+			case v1.ContainerFunctionNetworkPolicyRunner:
+				out.Network.Policy = fnv1alpha1.NetworkPolicy_NETWORK_POLICY_RUNNER
+			}
+		}
+	}
+	return out
+}
+
+// An UndesiredComposedResourceDeleter deletes composed resources from the API
+// server and from Composition state if their state doesn't include a FunctionIO
+// desired resource. This indicates the composed resource either exists or was
+// going to be created by P&T composition, but didn't survive the Composition
+// Function pipeline.
+type UndesiredComposedResourceDeleter struct {
+	client client.Writer
+}
+
+// NewUndesiredComposedResourceDeleter returns a ComposedResourceDeleter that
+// deletes undesired composed resources from both the API server and Composition
+// state.
+func NewUndesiredComposedResourceDeleter(c client.Writer) *UndesiredComposedResourceDeleter {
+	return &UndesiredComposedResourceDeleter{client: c}
+}
+
+// DeleteComposedResources deletes any composed resource that didn't come out the other
+// end of the Composition Function pipeline (i.e. that wasn't in the final
+// desired state after running the pipeline). Composed resources are deleted
+// from both the supposed composition state and from the API server.
+func (d *UndesiredComposedResourceDeleter) DeleteComposedResources(ctx context.Context, s *PTFCompositionState) error {
+	for name, cd := range s.ComposedResources {
+		// We know this resource is still desired because we recorded its
+		// desired state after running the FunctionIO pipeline. Don't garbage
+		// collect it.
+		if cd.Desired != nil {
+			continue
+		}
+
+		// We don't desire this resource to exist; remove it from our state.
+		delete(s.ComposedResources, name)
+
+		// No need to garbage collect resources that don't exist.
+		if !meta.WasCreated(cd.Resource) {
+			continue
+		}
+
+		// We want to garbage collect this resource, but we don't control it.
+		if c := metav1.GetControllerOf(cd.Resource); c == nil || c.UID != s.Composite.GetUID() {
+			continue
+		}
+
+		if err := d.client.Delete(ctx, cd.Resource); resource.IgnoreNotFound(err) != nil {
+			return errors.Wrapf(err, errFmtDeleteCD, cd.ResourceName, cd.Resource.GetObjectKind().GroupVersionKind().Kind, cd.Resource.GetName())
+		}
+	}
+
+	return nil
+}
+
+// UpdateResourceRefs updates the supplied state to ensure the XR references all
+// composed resources that exist or are pending creation.
+func UpdateResourceRefs(s *PTFCompositionState) {
+	refs := make([]corev1.ObjectReference, 0, len(s.ComposedResources))
+	for _, cd := range s.ComposedResources {
+		// Don't record references to resources that don't exist and that failed
+		// to render. We won't apply (i.e. create) these resources this time
+		// around, so there's no need to create dangling references to them.
+		if !meta.WasCreated(cd.Resource) && !cd.Rendered {
+			continue
+		}
+		ref := meta.ReferenceTo(cd.Resource, cd.Resource.GetObjectKind().GroupVersionKind())
```

Comment:
Yeah, this is because we're trying to deprecate the use of `corev1.ObjectReference` in general (iirc it's deprecated upstream). In practice our resourceRefs are currently of that type though so somewhat unavoidable.